nimbus/src/lib.rs

```diff
@@ -37,12 +37,21 @@ use uuid::Uuid;
 const DEFAULT_TOTAL_BUCKETS: u32 = 10000;
 const DB_KEY_NIMBUS_ID: &str = "nimbus-id";
 
+// The main `NimbusClient` struct must not expose any methods that make an `&mut self`,
+// in order to be compatible with the uniffi `[Threadsafe]` annotation. This is a helper
+// struct to contain the bits that do actually need to be mutable, so they can be
+// protected by a Mutex.
+#[derive(Default)]
+struct InternalMutableState {
+    available_randomization_units: AvailableRandomizationUnits,
+}
+
```

Comment:
This `InternalMutableState` stuff is new since #80, designed to make this work with the new requiements that we be `Sync` and `Send`. I've opted to create an obvious place for future additions of mutable state rather than risk a proliferation of individual mutexes.

Reply:
It makes me slightly uncomfortable we have 4 mutexes now - I assume you considered refactoring the database cache into the new mutable state abstraction but decided against it for some reason? I don't have any concrete suggestions here though, so it's not going to block an r+...

Reply:
Ref https://github.com/mozilla/nimbus-sdk/pull/94 for a followup that removes one of the mutexes.

Actually I misread your comment as being about the database rather than the database cache; I'll also think about whether it makes sense to move the dbcache into here as well, good suggestion!